signer/tests/integration/rbf.rs

```diff
@@ -0,0 +1,277 @@
+            .collect(),
+        signer_state: SignerBtcState {
+            utxo: SignerUtxo {
+                outpoint: OutPoint::new(signer_utxo.txid, signer_utxo.vout),
+                amount: signer_utxo.amount.to_sat(),
+                public_key: signers_public_key,
+            },
+            fee_rate: ctx.initial_fee_rate,
+            public_key: signers_public_key,
+            last_fees: None,
+        },
+        accept_threshold: 4,
+        num_signers: 7,
+    };
+
+    // Okay, lets submit the transaction. We also do a sanity check where
+    // we try to submit an RBF transaction with an insufficient fee bump.
+    // We need to note the fee for original transaction, so it is returned.
+    let (last_fee, last_fee_rate) = {
+        // There should only be one transaction here since there is only one
+        // deposit request and no withdrawal requests.
+        let mut transactions: Vec<UnsignedTransaction> = requests.construct_transactions().unwrap();
+        assert_eq!(transactions.len(), 1);
+        let mut unsigned: UnsignedTransaction = transactions.pop().unwrap();
+
+        let last_fee: u64 = unsigned.input_amounts() - unsigned.output_amounts();
+
+        // Add the signature and/or other required information to the witness data.
+        regtest::set_witness_data(&mut unsigned, signer.keypair);
+
+        rpc.send_raw_transaction(&unsigned.tx).unwrap();
+
+        // Step 2: create an RBF transaction that will fail.
+        //
+        // This is a little sanity check where we submit an RBF transaction
+        // but where we change the fee but an amount that is too small.
+        let mut transactions = requests.construct_transactions().unwrap();
+        let mut unsigned = transactions.pop().unwrap();
+
+        // We increase the fee paid but not by enough to be accepted
+        unsigned.tx.output[0].value -= Amount::from_sat(10);
+
+        regtest::set_witness_data(&mut unsigned, signer.keypair);
+
+        match rpc.send_raw_transaction(&unsigned.tx) {
+            Err(BtcRpcError::JsonRpc(JsonRpcError::Rpc(RpcError { message, .. }))) => {
+                assert!(message.starts_with("insufficient fee, rejecting replacement"))
+            }
+            _ => panic!("Unexpected response when sending bad replacement transaction"),
+        }
+        (last_fee, last_fee as f64 / unsigned.tx.vsize() as f64)
+    };
+
+    // Step 3. Construct an RBF transaction
+    //
+    // Let's update the request state with the new fee rate, the last fee amount paid
+    // and modify the outstanding deposits and withdrawals.
+    requests.signer_state.fee_rate = ctx.rbf_fee_rate;
+    requests.signer_state.last_fees = Some(Fees {
+        total: last_fee,
+        rate: last_fee_rate,
+    });
+
+    requests.deposits = deposits.clone();
+    requests.withdrawals = withdrawals.clone();
+    requests.deposits.truncate(ctx.rbf_deposits);
+    requests.withdrawals.truncate(ctx.rbf_withdrawals);
```

Comment:
This seems to be the only place `requests` is mutated. It would be easier to read and reason about the code if we could declare `requests` as immutable and move the mutation to a separate function, and just re-assign a new immutable `requests` variable here:

Something along the lines of
```suggestion
    let requests = update_request_state(&ctx, deposits, withdrawals, requests);
```

Reply:
Sure thing.

Reply:
Thank you! :pray: